Python/pystate.c

```diff
@@ -628,105 +731,163 @@ allocate_chunk(int size_in_bytes, _PyStackChunk* previous)
     return res;
 }
 
-static PyThreadState *
-new_threadstate(PyInterpreterState *interp, int init)
+static void
+init_threadstate_static_data(PyThreadState *tstate, PyThreadState *other)
 {
-    _PyRuntimeState *runtime = interp->runtime;
-    PyThreadState *tstate = (PyThreadState *)PyMem_RawMalloc(sizeof(PyThreadState));
-    if (tstate == NULL) {
-        return NULL;
+    assert(other != NULL);
+    assert(other != tstate);
+    assert(other->_preallocated.initialized);
+    assert(!tstate->_preallocated.initialized);
+
+    // We only want to copy static data that every thread state has in common.
+    // Hence, if "other" was statically initialized
+    // then we only want that data.
+
+    /* Initialize tstate->_preallocated. */
+    tstate->_preallocated.initialized = 1;
+}
+
+/* Fully initialize the (low-level) thread state.
+   Other (high-level) init happens in pylifecycle.c.
+   If this is a re-used thread state then it will be zeroed out first.
+   Otherwise all fields not initialized here are expected to be zeroed out,
+   e.g. by PyMem_RawCalloc() or memset().
+   The interpreter is not manipulated.  Instead it is assumed that
+   the thread state is getting added to the interpreter.
+  */
+
+static void
+init_threadstate(PyThreadState *tstate,
```

Comment:
This should be error if called with an already initialized thread, IMO.
What's the justification for messing with a thread that is already in use? It seems unsafe to me.

Reply:
When `Py_Initialize()` is run a second time (after `Py_Finalize()`) we re-use the `_PyRuntime` static global.  Consequently we re-use its static main interpreter and that interpreter's static thread state.  That is the reason.  I agree the check there isn't perfect, but it isn't so much a problem since this function isn't part of the API.